### ISSUE TITLE
Array module review followup: Rename Array front/back to fist/last

### ISSUE THREAD

**Issue by stonea** [chapel-lang/chapel#18087]
Array module review followup: Rename Array front/back to fist/last

This is a followup item for the [Array module review](https://github.com/Cray/chapel-private/issues/2241).

One item we had agreement on is that we should rename the array front and back methods to first and last.  This provides consistency with domain which uses first and last to refer to the first and last indices.

See also this comment:

https://github.com/Cray/chapel-private/issues/2194#issuecomment-870947834

**Comment by bradcray** [comment]
I'm a _huge_ fan of making this change.

**Comment by e-kayrakli** [comment]
I'll just add here that while doing the range re-review, we were leaning towards keeping `range.first`, `range.last` paren-less. Array's `first` and `last` should be parenless, as well. (Or if we gravitate towards parenful, we should use parenful `first`/`last` everywhere)

**Comment by bradcray** [comment]
I mostly agree with Engin's comment in the previous bullet, though I was trying to see whether there were limits that would break that feeling for me.  The best I could come up with "What if I had a singly-linked list data structure that only stores 'head' and not 'tail'...  Would I be willing to support `last()` without parens given that it's an O(n) operation?"  I think the answer is "no", but also that none of our standard libraries should support an O(n) `last` operation.  So I also don't think I'd insist that the library routines we do support should support paren-ful first/last in case someone did create such a case down the road...

**Comment by e-kayrakli** [comment]
Would you feel safer and better if we had all those builtin types have parenful `first`/`last`?

I don't have a strong preference, but left to my own devices would make these parenless, which I think is what you're leaning towards even after the thought experiment above?

**Comment by bradcray** [comment]
That's right, I still lean toward paren-less, though if you made me kick two 0-argument range queries into the paren-ful camp, those would be the ones.

And then just to make clear how on the fence I still am...  `myRange.first` and `myDenseRectangularDomain.first` both seem eminently reasonable to me, but `myDenseRectangularArray.first` does feel slightly odder...  perhaps because because array accesses usually involve `[]`/`()`?  Or because I don't think of the first element of an array as being a particularly innate property of the array?  So if others agreed and thought this was a reason to make them all paren-ful, I could get behind that.  Or if the decision was to make cases where it was more like an innate property paren-less and those where it was more like a collection lookup paren-ful, I could get behind that as well.

**Comment by stonea** [comment]
I have a PR ready to go to implement this change: https://github.com/chapel-lang/chapel/pull/18256

But it sounds like maybe there's some bigger questions about parenless vs. parenful methods for arrays, ranges, and domains. I've opened an issue for that discussion: https://github.com/Cray/chapel-private/issues/2419 . I'll keep the PR open without submitting it until we arrive at consensus.
